File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Die/KeepObjectDie.cpp
// FILE: KeepObjectDie.cpp ///////////////////////////////////////////////////////////////////////////
// Author: Kris Morness, November 2002
// Desc:   Die module for things that want to leave rubble in the world and don't have other die
//         modules. This fixes civilian buildings that don't have garrison contains. Garrison
//         contains have a die module built in, so these buildings need something. Without it
//         they default to the destroydie module which outright removes the object.
///////////////////////////////////////////////////////////////////////////////////////////////////

// INCLUDES ///////////////////////////////////////////////////////////////////////////////////////
#include "PreRTS.h"	// This must go first in EVERY cpp file int the GameEngine

#include "Common/Xfer.h"
#include "GameClient/Drawable.h"
#include "GameLogic/GameLogic.h"
#include "GameLogic/Module/KeepObjectDie.h"

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
KeepObjectDie::KeepObjectDie( Thing *thing, const ModuleData* moduleData ) : DieModule( thing, moduleData )
{
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
KeepObjectDie::~KeepObjectDie( void )
{
}

//-------------------------------------------------------------------------------------------------
/** The die callback. */
//-------------------------------------------------------------------------------------------------
void KeepObjectDie::onDie( const DamageInfo *damageInfo )
{
	if( !isDieApplicable(damageInfo) )
	{
		return;
	}
}

// ------------------------------------------------------------------------------------------------
/** CRC */
// ------------------------------------------------------------------------------------------------
void KeepObjectDie::crc( Xfer *xfer )
{

	// extend base class
	DieModule::crc( xfer );

}  // end crc

// ------------------------------------------------------------------------------------------------
/** Xfer method
	* Version Info:
	* 1: Initial version */
// ------------------------------------------------------------------------------------------------
void KeepObjectDie::xfer( Xfer *xfer )
{

	// version
	XferVersion currentVersion = 1;
	XferVersion version = currentVersion;
	xfer->xferVersion( &version, currentVersion );

	// extend base class
	DieModule::xfer( xfer );

}  // end xfer

// ------------------------------------------------------------------------------------------------
/** Load post process */
// ------------------------------------------------------------------------------------------------
void KeepObjectDie::loadPostProcess( void )
{

	// extend base class
	DieModule::loadPostProcess();

}  // end loadPostProcess

File: GeneralsRust/Code/GameEngine/GameLogic/src/object/die/keep_object_die.rs
//! KeepObjectDie - Die module that leaves the object as wreckage
//!
//! Original C++ location: GameLogic/Module/KeepObjectDie.h/.cpp
//! Original C++ Author: Kris Morness, November 2002
//! Rust conversion: 2025

use super::{DieModule, DieModuleData, DieModuleInterface};
use crate::damage::DamageInfo;
use crate::object::Object;
use std::sync::{Arc, RwLock};

/// KeepObjectDie - Keeps the object in the world as a corpse/wreckage
///
/// This die module does not remove the object from the game world.
/// Instead, it leaves it in place as wreckage or a corpse. This is used
/// for civilian buildings that don't have garrison contains, allowing them
/// to leave rubble without the DestroyDie module removing them entirely.
/// (Matches C++ KeepObjectDie)
#[derive(Debug)]
pub struct KeepObjectDie {
    base: DieModule<DieModuleData>,
}

impl KeepObjectDie {
    /// Create a new KeepObjectDie module
    pub fn new(object: Arc<RwLock<Object>>, module_data: Arc<DieModuleData>) -> Self {
        Self {
            base: DieModule::new(object, module_data),
        }
    }

    /// Get module name
    pub fn get_module_name() -> &'static str {
        "KeepObjectDie"
    }
}

impl DieModuleInterface for KeepObjectDie {
    /// Called when the object dies - keeps the object in place
    /// (Matches C++ KeepObjectDie::onDie)
    fn on_die(&mut self, object: &mut Object, damage_info: &DamageInfo) {
        // Check if this die module should activate
        if !self.is_die_applicable(object, damage_info, &self.base.module_data.die_mux_data) {
            return;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_keep_object_die_module_name() {
        assert_eq!(KeepObjectDie::get_module_name(), "KeepObjectDie");
    }
}
